### DISCUSSION TITLE
How does one mass execute a migration-like callback on all tenants to remove a feature?

### DISCUSSION THREAD

**Discussion by willnationsdev** [OrchardCMS/OrchardCore#16468] [Q&A] [answered]
I have the `OrchardCore.Contents.FileContentDefinitions` feature turned on. I want to mass disable this feature and migrate the file data into the database. I thought about using `IFeatureEventHandler` to use the `DisabledAsync` callback for when a tenant's `OrchardCore.Contents.FileContentDefinitions` feature is disabled to actually move the data; however, the `OrchardCoreBuilder` API only seems to offer the `EnableFeature("...")` method without any disabling equivalent.

I also tried registering a Migrations class into my main application, thinking that it would run in all tenants and that I could use the `IShellFeaturesManager` in that migration class to disable it everywhere, but the migration never triggered.

Thank you

**Comment by Piedone** [comment, marked as answer]
You can disable the feature with `IFeatureManager` (and run any other code) on all tenants, run from the Default tenant, like this: https://orcharddojo.net/blog/how-to-access-services-from-another-tenant-in-orchard-core-orchard-nuggets

**Reply by willnationsdev** [reply]
Thank you @Piedone. I've seen this code snippet before, and in principle it sounds like a good solution, but where would I *put* this logic in order for it to perform a one-time update? I don't want it in an arbitrary Controller endpoint. If I have to put it in some kind of event handler service or migration class, it's always tied to some module that I would then have to explicitly enable in the Default tenant in order to get it to load that logic (which seems unnecessary). I don't need my desire to disable the feature to itself be some feature that needs enabling.

Alternatively, could I add a step in the bad feature's DataMigrations class that makes it disable *itself*? Would that even work?

**Reply by Piedone** [reply]
Well, if you want to run the code, you need have that code enabled :). But better to have to do a single time on the Default tenant than on all the tenants. Especially for a one-time update I don't see how even a controller would be a bad choice, since it doesn't matter.

You can disable the feature from a migration of its, yes, but for that, you'd need to modify an OC module.
